oauth-server/src/main/java/io/strimzi/kafka/oauth/server/OAuthKafkaPrincipalBuilder.java

```diff
@@ -0,0 +1,134 @@
+ * the OAuthBearerToken token produced by <em>io.strimzi.kafka.oauth.server.JaasServerOauthValidatorCallbackHandler</em>.
+ * </p>
+ * <p>
+ * You can use 'principal.builder.class=io.strimzi.kafka.oauth.server.OAuthKafkaPrincipalBuilder'
+ * property definition in server.properties to install it.
+ * </p>
+ */
+public class OAuthKafkaPrincipalBuilder extends DefaultKafkaPrincipalBuilder implements Configurable {
+
+    private static final SetAccessibleAction SET_PRINCIPAL_MAPPER = SetAccessibleAction.newInstance();
+
+    private static class SetAccessibleAction implements PrivilegedAction<Void> {
+
+        private Field field;
+
+        SetAccessibleAction(Field field) {
+            this.field = field;
+        }
+
+        @Override
+        public Void run() {
+            field.setAccessible(true);
+            return null;
+        }
+
+        void invoke(DefaultKafkaPrincipalBuilder target, Object value) throws IllegalAccessException {
+            AccessController.doPrivileged(this);
+            field.set(target, value);
+        }
+
+        static SetAccessibleAction newInstance() {
+            try {
+                return new SetAccessibleAction(DefaultKafkaPrincipalBuilder.class.getDeclaredField("sslPrincipalMapper"));
+            } catch (NoSuchFieldException e) {
+                throw new IllegalStateException("Failed to install OAuthKafkaPrincipalBuilder. This Kafka version does not seem to be supported", e);
+            }
+        }
+    }
+
+
+    public OAuthKafkaPrincipalBuilder() {
+        super(null, null);
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+
+        Object sslPrincipalMappingRules = configs.get(BrokerSecurityConfigs.SSL_PRINCIPAL_MAPPING_RULES_CONFIG);
+        Object sslPrincipalMapper;
+
+        try {
+            Class<?> clazz = Class.forName("org.apache.kafka.common.security.ssl.SslPrincipalMapper");
+            try {
+                Method m = clazz.getMethod("fromRules", List.class);
+                if (sslPrincipalMappingRules == null) {
+                    sslPrincipalMappingRules = Collections.singletonList("DEFAULT");
+                }
+                sslPrincipalMapper = m.invoke(null, sslPrincipalMappingRules);
+
+            } catch (NoSuchMethodException ex) {
+                Method m = clazz.getMethod("fromRules", String.class);
+                if (sslPrincipalMappingRules == null) {
+                    sslPrincipalMappingRules = "DEFAULT";
+                }
+                sslPrincipalMapper = m.invoke(null, sslPrincipalMappingRules);
+            }
+
+            // Hack setting sslPrincipalMapper to DefaultKafkaPrincipalBuilder
+            // An alternative would be to copy paste the complete DefaultKafkaPrincipalBuilder implementation
+            // into this class and extend it
+
+            SET_PRINCIPAL_MAPPER.invoke(this, sslPrincipalMapper);
```

Comment:
This bit is horrible though. Can't you just instantiate your own `SslPrincipalMapper` using the `configs` and not subclass `DefaultKafkaPrincipalBuilder` at all. This looks like it could just break with a new upstream Kafka which made changes in this area.

Reply:
Maybe, but I'd leave this for another PR.